network/dhcpd.py

```diff
@@ -0,0 +1,189 @@
+    HAVE_PYPUREOMAPI=True
+except ImportError:
+    HAVE_PYPUREOMAPI=False
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            state   = dict(default='present', choices=['present', 'absent']),
+            gather_facts = dict(default=None, choices=BOOLEANS),
+            mac_address = dict(default=None, type='str'),
+            ip_address = dict(default=None, type='str'),
+            server = dict(required=True, type='str'),
+            port = dict(default=7911, type='int'),
+            key_name = dict(default=None, type='str'),
+            key_secret = dict(default=None, type='str')
+        ),
+        required_together = [
+                                ['key_name', 'key_secret']
+                            ],
+        required_one_of = [
+                                ['mac_address', 'ip_address'],
+                                ['gather_facts', 'state']
+                          ],
+        supports_check_mode=False
+    )
+
+    if not HAVE_PYPUREOMAPI:
+        module.fail_json(msg="The pypureomapi python library is required")
+
+    if (module.params['state'] is None) or (module.params['state'] == 'absent'):
+        if not ( (module.params['mac_address'] is not None) ^ (module.params['ip_address'] is not None) ):
+            module.fail_json(msg="state=absent or gather_facts=true require exactly either one Hardware or one IP address")
+    if module.params['state'] == 'present':
+        if not ( (module.params['mac_address'] is not None) and (module.params['ip_address'] is not None) ):
+            module.fail_json(msg="state=present requires both an IP and a Hardware address")
+
+    try:
+        dhcpd_omapi = pypureomapi.Omapi(module.params['server'], module.params['port'], module.params['key_name'], module.params['key_secret'])
+    except pypureomapi.OmapiError:
+        module.fail_json(msg="Error connecting to dhcp server")
+
+    #if state is none then query dhcpd and return as facts
+    if module.boolean(module.params['gather_facts']):
+        ansible_facts_dict = {
+            "changed" : False,
+            "ansible_facts": {
+                }
+        }
+        if module.params['ip_address'] is not None:
+            try:
+                mac = dhcpd_omapi.lookup_mac(module.params['ip_address'])
+                ansible_facts_dict['ansible_facts']['dhcp_ip'] = module.params['ip_address']
+                ansible_facts_dict['ansible_facts']['dhcp_mac'] = mac
+                print json.dumps(ansible_facts_dict)
+                return 0
+            except pypureomapi.OmapiErrorNotFound:
+                module.fail_json(msg="Entry not found")
+        else:
+            try:
+                ip = dhcpd_omapi.lookup_ip(module.params['mac_address'])
+                ansible_facts_dict['ansible_facts']['dhcp_ip'] = ip
+                ansible_facts_dict['ansible_facts']['dhcp_mac'] = module.params['mac_address']
+                print json.dumps(ansible_facts_dict)
```

Comment:
facts should be handled by a facts module, we separate them from management modules